mllib/src/main/scala/org/apache/spark/ml/feature/VarianceThresholdSelector.scala

```diff
@@ -0,0 +1,278 @@
+ * limitations under the License.
+ */
+
+package org.apache.spark.ml.feature
+
+import scala.collection.mutable.ArrayBuilder
+
+import org.apache.hadoop.fs.Path
+
+import org.apache.spark.annotation.Since
+import org.apache.spark.ml._
+import org.apache.spark.ml.attribute.{Attribute, AttributeGroup, NominalAttribute}
+import org.apache.spark.ml.linalg._
+import org.apache.spark.ml.param._
+import org.apache.spark.ml.param.shared._
+import org.apache.spark.ml.stat.Summarizer
+import org.apache.spark.ml.util._
+import org.apache.spark.sql._
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types.{StructField, StructType}
+
+
+/**
+ * Params for [[VarianceThresholdSelector]] and [[VarianceThresholdSelectorModel]].
+ */
+private[feature] trait VarianceThresholdSelectorParams extends Params
+  with HasFeaturesCol with HasOutputCol {
+
+  /**
+   * Param for variance threshold. Features with a variance lower than this threshold will
+   * be removed.
+   *
+   * @group param
+   */
+  @Since("3.1.0")
+  final val varianceThreshold = new DoubleParam(this, "varianceThreshold",
+    "Param for variance threshold. Features with a variance lower than this threshold" +
+      " will be removed.", ParamValidators.gt(0))
+
+  /** @group getParam */
+  @Since("3.1.0")
+  def getVarianceThreshold: Double = $(varianceThreshold)
+}
+
+/**
+ * Feature selector that removes all low-variance features. Features with a
+ * variance lower than the threshold will be removed. The default is to keep
+ * all features with non-zero variance, i.e. remove the features that have the
+ * same value in all samples.
+ */
+@Since("3.1.0")
+final class VarianceThresholdSelector @Since("3.1.0")(@Since("3.1.0") override val uid: String)
+  extends Estimator[VarianceThresholdSelectorModel] with VarianceThresholdSelectorParams
+with DefaultParamsWritable {
+
+  @Since("3.1.0")
+  def this() = this(Identifiable.randomUID("VarianceThresholdSelector"))
+
+  /** @group setParam */
+  @Since("3.1.0")
+  def setVarianceThreshold(value: Double): this.type = set(varianceThreshold, value)
+
+  /** @group setParam */
+  @Since("3.1.0")
+  def setFeaturesCol(value: String): this.type = set(featuresCol, value)
+
+  /** @group setParam */
+  @Since("3.1.0")
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  @Since("3.1.0")
+  override def fit(dataset: Dataset[_]): VarianceThresholdSelectorModel = {
+    transformSchema(dataset.schema, logging = true)
+
+    val Row(maxs: Vector, mins: Vector, variances: Vector) = dataset
+      .select(Summarizer.metrics("max", "min", "variance").summary(col($(featuresCol)))
+        .as("summary"))
+      .select("summary.max", "summary.min", "summary.variance")
+      .first()
+
+    val result = variances.toArray.zip(maxs.toArray).zip(mins.toArray).zipWithIndex
+    // if varianceThreshold not set, remove the features that have the same value in all samples.
+    val features = if (!isSet(varianceThreshold)) {
```

Comment:
Why not giving `varianceThreshold` a default value = 0?

Reply:
I thought we keep the features >= threshold, so I can't default it to 0 (variance 0 will be kept too). 
Maybe I should change the definition to "keep the features > threshold"?

Reply:
I will change to "keep the features > threshold". I looked sklearn code and did a little test, sklearn removes the <= threshold features. 
```
    def _get_support_mask(self):
        check_is_fitted(self)

        return self.variances_ > self.threshold
```

```
>>> data = [[0, 2, 7, 1],
...         [1, 4, 9, 5]]
>>> np.var([0, 1])
0.25
>>> np.var([2, 4])
1.0
>>> np.var([7, 9])
1.0
>>> np.var([1, 5])
4.0
>>> VarianceThreshold(threshold=0.2499).fit_transform(data)
array([[0, 2, 7, 1],
       [1, 4, 9, 5]])
>>> VarianceThreshold(threshold=0.25).fit_transform(data)
array([[2, 7, 1],
       [4, 9, 5]])
>>> VarianceThreshold(threshold=0.9999).fit_transform(data)
array([[2, 7, 1],
       [4, 9, 5]])
>>> VarianceThreshold(threshold=1.0).fit_transform(data)
array([[1],
       [5]])
```